internal/controller/apiextensions/conditions/conditions.go

```diff
@@ -0,0 +1,224 @@
+	// TODO(dalton): test that this does not update the ConditionProcessor
+	for _, o := range opts {
+		o(&cp)
+	}
+	prev := GetConditions(cp.Obj)
+
+	cmp := prev.Compare(cm)
+	updated := len(cmp.New) > 0
+	if updated {
+		RecordConditions(cp.Obj, cp.Recorder, cmp.New)
+	}
+
+	var nm ConditionMap
+	if cp.DropStale {
+		nm = mergeMaps(cmp.Equal, cmp.New)
+		if len(cmp.Old) > 0 {
+			// old entries were dropped
+			updated = true
+		}
+	} else {
+		nm = mergeMaps(cmp.Old, cmp.Equal, cmp.New)
+	}
+
+	// if the nm does not contain a system condition, try to copy that condition
+	// from the previous conditions
+	for k := range systemConditionTypes {
+		if _, ok := nm[k]; ok {
+			continue
+		}
+		if c, ok := prev[k]; ok {
+			nm[k] = c
+		}
+	}
+
+	if updated {
+		// only update if a change occurred to prevent needless reconciliations
+		ForceSetConditions(cp.Obj, nm)
+	}
+}
+
+func ForceSetConditions(u *unstructured.Unstructured, cm ConditionMap) {
+	conditioned := xpv1.ConditionedStatus{}
+	for _, c := range cm {
+		conditioned.Conditions = append(conditioned.Conditions, c.Condition)
+	}
+	_ = fieldpath.Pave(u.Object).SetValue("status.conditions", conditioned.Conditions)
+}
+
+type CompareResult struct {
+	New   ConditionMap
+	Equal ConditionMap
+	Old   ConditionMap
+}
+
+func (cm ConditionMap) Compare(other ConditionMap) CompareResult {
+	res := CompareResult{
+		New:   make(ConditionMap),
+		Equal: make(ConditionMap),
+		Old:   make(ConditionMap),
+	}
+	for k, a := range cm {
+		if b, ok := other[k]; ok && a.Equal(b.Condition) {
+			res.Equal[k] = a
+		} else {
+			res.Old[k] = a
+		}
+	}
+	for k, b := range other {
+		if a, ok := cm[k]; !ok || !b.Equal(a.Condition) {
+			res.New[k] = b
+		}
+	}
+	return res
+}
+
+func RecordConditions(obj runtime.Object, r event.Recorder, cm ConditionMap) {
+	for _, c := range cm {
+		switch c.EventType {
+		case event.TypeWarning:
+			r.Event(obj, event.Warning(event.Reason(c.Reason), errors.New(c.Message)))
+		case event.TypeNormal:
+			r.Event(obj, event.Normal(event.Reason(c.Reason), c.Message))
+		}
+	}
+}
+
+// SetClaimConditionTypes sets the composite's status.claimConditions to whatever is provided
+// as the conditions arg. All existing conditions will be wiped.
+func SetClaimConditionTypes(xr *composite.Unstructured, cm ConditionMap) {
+	ts := []xpv1.ConditionType{}
+	for t := range cm {
+		ts = append(ts, t)
+	}
+	_ = fieldpath.Pave(xr.Object).SetValue("status.claimConditions", ts)
+}
+
+// GetConditions returns all items from the object's status.conditions.
+func GetConditions(u *unstructured.Unstructured) ConditionMap {
+	conditioned := xpv1.ConditionedStatus{}
+	// The path is directly `status` because conditions are inline.
+	if err := fieldpath.Pave(u.Object).GetValueInto("status", &conditioned); err != nil {
+		return make(ConditionMap)
+	}
+	m := make(ConditionMap)
+	for _, c := range conditioned.Conditions {
+		m[c.Type] = Condition{
+			Condition: c,
+		}
+	}
+	return m
+}
```

Comment:
These look similar to e.g. https://github.com/crossplane/crossplane-runtime/blob/master/pkg/resource/unstructured/composite/composite.go#L207